### DISCUSSION TITLE
JsonScada Dnp3

### DISCUSSION THREAD

**Discussion by rudyroc** [riclolsen/json-scada#25] [Q&A] [answered]
Hi Ricardo
I would appreciate a lot  your help about Dnp3 implementation in JsonScada. Few days ago I started with JsonScada, I have installed it in Linux Debian 10. I followed the instructions detailed at Readme. 
I have configurated one Dnp3 instance in  the Admin panel, but when the Dnp3 process is started it's gives an error (screenshot # 1). I tried to start the Dnp3Client located at /json-scada/bin using ./Dnp3Client 1 1 with the same result. (Screnshot # 2).
After that,I tried JsonScada Windows version 0.8. It worked  correctly with 104 Client & Server, but when I enabled the Dnp3Client service it's fails to start ( screenshot # 3 ).
Am I making a mistake? what am I doing wrong?
Best regards,
Rudy
![screenshot_#_1](https://user-images.githubusercontent.com/84212412/118345112-d43c0080-b508-11eb-86a6-4952361d7c72.png)
![screenshot_#_2](https://user-images.githubusercontent.com/84212412/118345113-d56d2d80-b508-11eb-999c-65c173e92443.png)
![screenshot_#_3](https://user-images.githubusercontent.com/84212412/118345115-d56d2d80-b508-11eb-8a37-a19eaee36243.png)

**Comment by riclolsen** [comment, marked as answer]
Hi Rudy.

I have found a problem of scope (garbage collection) with the DNP3 manager object.
You can update the code and try to rebuild the driver on Linux.

Or use this updated driver on Windows (extract to c:\json-scada\bin\). 

[dnp3client-0.1.1.zip](https://github.com/riclolsen/json-scada/files/6487594/dnp3client-0.1.1.zip)

**Comment by riclolsen** [comment]
Dear Rudy,
If possible, please report your results using this driver. I do not have a production use case for it at the moment. So, any feedback is important for the project.
Thank you very much for spotting this problem.
Best regards,
Ricardo

**Comment by rudyroc** [comment]
Hi Ricardo
Thank you for the quick response. I'll test the modifications and let you know the result.
Best Regards,
Rudy

**Comment by rudyroc** [comment]
Dear Ricardo,
Bad news, I  tested the new windows version of Dnp3Client and I got the same result ( Screenshot #10).
The same thing happened on Linux.
Best regards,
Rudy
![screenshot_#_10](https://user-images.githubusercontent.com/84212412/118382946-3f9cd580-b5d0-11eb-95b4-bff5314d8874.png)

**Reply by riclolsen** [reply]
Probably open ssl is missing.
Make sure Windows is 64 bits.
Install Open SSL binaries for 64-bit Windows, https://slproweb.com/download/Win64OpenSSL_Light-1_1_1k.msi.
Install https://dotnet.microsoft.com/download/dotnet/thank-you/runtime-5.0.6-windows-x64-installer
Reboot and run dnp3client.exe again.
If the problem persists, replace the bin folder with this new full binary build
https://drive.google.com/file/d/1comxhyEdrILclM92_AyHbxTSC4PrgoQq/view?usp=sharing

On Linux please check if iec104client also crashes. Check if you have dot core sdk 5.0. I have yet to test on Linux.

**Reply by riclolsen** [reply]
Testing on Linux was successful for DNP3.

**Comment by riclolsen** [comment]
Windows installer release 0.9 is out. It solves issues with DNP3 and has no dependency of openssl installation.
Check it out.

**Comment by rudyroc** [comment]
Dear Ricardo,
Thank you for the new version. Unfortunately, last week I couldn't test the suggested changes.
This weekend I'll try the new windows versions and I will try to rebuild the linux version and let you know the result.
 Best regards,
Rudy

**Comment by rudyroc** [comment]
Hi Ricardo
I have installed the version v 0.10 and i did a very simple configuration.
Four di's : id 1000 .......1003
Two ai's: id 60000....60001
One do: id 90000
One calculated point: id 100
Dnp3 is the only enabled protocol and I disabled demo service.
Dnp3 service start successfully, but when one or more di's changes the origin timestamp isn't recorded into de database.
I monitored the communication, using wireshark, and the information exchange looks ok, but the timestamp in alarm screen is the PC time and the event screen shows a bad time.On the other hand,only the changes from off to on are showed. I attach some screenshots about that.
Analog changes are showed at event's screen with a timestamp wich differs from the protocol timestamp.
About calculated point, I used the formula 54 with the points 1000 and 1001 as parcells, but it doesn't work.
I assume that the malfunction is due to errors in the configuration done, so I would appreciate your help in correcting the errors.
I attach some screenshots and database collections with configuration and events.
Best regards,
Rudy
[test_json.zip](https://github.com/riclolsen/json-scada/files/6690876/test_json.zip)
[screenshots.zip](https://github.com/riclolsen/json-scada/files/6690878/screenshots.zip)

**Reply by riclolsen** [reply]
Hi Rudy!

The screenshots show problems with timestamps, time apparently is zeroed however valueStringAtSource is correct. I will check this probable bug and provide a fix ASAP.

For digital tags, change the "isEvent": true, to "isEvent": false or only off->on changes will be updated (as you reported).

"tag": "CVTGSACAINT._3L015", never updated. This tag has "protocolSourceCommonAddress": 0, you should configure this as 1 probably.
"tag": "CVTGSACAINT._3L015_CERRADO","CVTGSACAINT._3L015_ABIERTO", these were distributed with object 1, var 2 ("asduAtSource": "1 2") so there are no timestamps coming from the RTU.
"tag": "CVTGSACAINT._3L025_CERRADO", "CVTGSACAINT._3L025_ABIERTO", object 1 var 1, also have no timestamps.
"tag": "CVTGSACA3L025_P", "CVTGSACA3L025_Q",, object 30 var 2, also have no timestamps.

Thanks for the tests and comments.

**Reply by riclolsen** [reply]
I have fixed the timestamp problem, date was ok but time was being discarded. Now it should update timestamps correctly.
Please replace these files in c:\json-scada\bin\ on Windows or rebuild from sources on Linux. 
[dnp3bin.zip](https://github.com/riclolsen/json-scada/files/6694568/dnp3bin.zip)

**Comment by rudyroc** [comment]
Hi Ricardo
Thank you very much for the corrections you have made. The timestamp works correctly for DI's.
Regard to analogs events, the timestamp is not recorded, but the response from the RTU is 32 Var 4. It would appear to be interpreted by the dnp driver as an object 30 Var 2. Attached is a screenshot of the dissection performed with Wireshark of the communication and another of the mongodb base.
In the DO's case I have configured a command point as Object 12 Var 1 type Trip / Close. The close command works correctly but the trip doesn't work. The RTU has the same configuration as that used with the OSHMI, where the trip /close commands worked well. I attach the Wireshark screenshots where you can compare the two cases (OSHMI/JSON).
Best regards,
Rudy
![json_dnp_analog_change_03072021](https://user-images.githubusercontent.com/84212412/124373528-50c9a080-dc69-11eb-94ba-5ba30761291d.png)
![mongodb_analog_1_03072021](https://user-images.githubusercontent.com/84212412/124373533-62ab4380-dc69-11eb-84d8-6be76ae684bf.png)
![json_dnp_close_03072021](https://user-images.githubusercontent.com/84212412/124373543-73f45000-dc69-11eb-9b5b-048a20d88db4.png)
![json_dnp_trip_03072021](https://user-images.githubusercontent.com/84212412/124373544-75257d00-dc69-11eb-8fc2-dd9941378355.png)
![oshmi_dnp_close_03072021](https://user-images.githubusercontent.com/84212412/124373545-75be1380-dc69-11eb-9cea-ee58a1cba0f1.png)
![oshmi_dnp_trip_03072021](https://user-images.githubusercontent.com/84212412/124373548-75be1380-dc69-11eb-9db6-05a8becaa329.png)

**Reply by riclolsen** [reply]
Regarding analog values with timestamps, I tested it and it worked from driver to mongo (sourceDataUpdate object), but it is not being carried out to the main record.

Driver output (run with parameters Dnp3Client.exe 1 2 to enable more log details) 

  04/07/2021 17:33:23 - DNP3 - <-AL- - FIR: 1 FIN: 1 CON: 1 UNS: 1 SEQ: 9 FUNC: UNSOLICITED_RESPONSE IIN: [0x00, 0x00]
  04/07/2021 17:33:23 - DNP3 - <-AL- - **032,004 Analog Input Event** - 16-bit With Flag and Time, 16-bit count and prefix [1]
  04/07/2021 17:33:23:604 DNP3: Analog, Ind: 0 value: 789 quality: 1 **time: 04/07/2021 17:33:20** [SYNCHRONIZED]
  04/07/2021 17:33:23 - DNP3 - -AL-> - FIR: 1 FIN: 1 CON: 0 UNS: 1 SEQ: 9 FUNC: CONFIRM
  04/07/2021 17:33:23:688 MongoDB - ADD - Connection: 2002 Group: 30 Address: 0 Value: 789
  
    "protocolSourceCommonAddress": 30,
    "protocolSourceConnectionNumber": 2002,
    "protocolSourceObjectAddress": 0,
    "sourceDataUpdate": {
      "valueAtSource": 789,
      "valueStringAtSource": "value: 789 quality: 1 time: 04/07/2021 17:33:20 [SYNCHRONIZED]",
      "valueJsonAtSource": { "Quality": {}, "Timestamp": {} },
      "asduAtSource": "32 4", <<<<<<<<< correct obj/var, yours is 30/2 so no timestamp
      "causeOfTransmissionAtSource": "3",
      "timeTagAtSource": { "$date": "2021-07-04T17:33:20.069Z" }, <<<<<<<<<<<< source timestamp ok
      "timeTagAtSourceOk": true,
      "timeTag": { "$date": "2021-07-04T20:33:23.604Z" },
      "notTopicalAtSource": false,
      "invalidAtSource": false,
      "overflowAtSource": false,
      "blockedAtSource": false,
      "substitutedAtSource": false,
      "carryAtSource": false,
      "transientAtSource": false,
      "originator": "DNP3|2002"
    },
    "stateTextFalse": "undefined",
    "stateTextTrue": "undefined",
    "substituted": false,
    "supervisedOfCommand": 0,
    "tag": "BRA-MG-MW",
    "timeTag": { "$date": "2021-07-04T20:33:24.074Z" },
    "timeTagAlarm": { "$date": "2021-07-04T20:21:25.929Z" },
    "timeTagAtSource": null, <<<<<<<<<<<<<<< should come to here but it is not happening for analog tags, to be fixed
    "timeTagAtSourceOk": null,
    "value": 789,
  
Probably the document from mongo you presented it is of a later scanned update without the timestamp.
Anyway I will take care of the forwarding problem.

**Reply by riclolsen** [reply]
Analog timestamps are fixed in the updated source code.
Update this file to fix the problem.
https://github.com/riclolsen/json-scada/blob/master/src/cs_data_processor/cs_data_processor.js

**Reply by riclolsen** [reply]
Indeed, JSON SCADA CROB controls could not replicate the mentioned OSHMI controls.

More options have been added to fix the problem. Also a fixed pulse time of 100ms was added.


Update the driver posted here and check new documentation
https://github.com/riclolsen/json-scada/blob/master/src/dnp3/Dnp3Client/README.md

Change this parameter to replicate the OHSMI control: protocolSourceCommandDuration=12.

This way both CLOSE and TRIP controls will have PULSE=ON with on and off times of 100ms.

[dnp3bin-1.1.2.zip](https://github.com/riclolsen/json-scada/files/6764949/dnp3bin-1.1.2.zip)

**Comment by rudyroc** [comment]
Hi Ricardo
Thanks for the quick response. I have tested the commands and these work correctly.
I have a doubt about analogs values with timestamp, ¿ these values should be logged whenever a change exists in the soeData collection ?.
I have simulated several changes of these values, the response from the RTU is always Object 32 Var 4 with timestamp, and sometimes they are recorded in the collection soeData and other times not. The value is always registered in the RealTimeData collection.
Best wishes,
Rudy

**Reply by riclolsen** [reply]
Good question, Rudy.
Analog changes with and without source timestamps are inserted into the "hist" postgresql table, it is also updated values on the realtimeData table. Postgresql/TimescaleDB is the best source for querying this type of data.
The soeData mongodb collection is intended for events, so not really for analog historical recording. This collection is defined as a capped collection with a limited (fixed) size. You can change the size or redefine it as a normal collection. I intend that only analog tags marked with isEvent=true be recorded here, but this is for measurements with sporadical changes. Measurements with continuous changes can overload this collection, reducing the performance of queries.

**Reply by rudyroc** [reply]
Hi Ricardo!
Thanks for the explanation. Now I'll build the Linux version and test it.
Any questions, I will consult you.
Best regards,
Rudy

**Comment by rudyroc** [comment]
Hi Ricardo,
I compiled version v.12 on linux. Everything works correctly except the Dnp3 driver.
I followed the instructions and installed all the necessary applications / dependencies.The installed dotnet version is "dotnet-sdk-5.0".
As you can see in the screenshot dnp3_1 all processes start up correctly, but the Dnp3Client process restarts continuously, the same image shows the dnp3 client log. Client error log is empty. I attach a copy of the supervisord.conf file.
 Finally, I tried to run the driver from a terminal;output can be seen in the screenshot dnp3_2.
Could this malfunction be due to a compilation error ?
Best regards,
Rudy
![dnp3_1](https://user-images.githubusercontent.com/84212412/126091958-6bdae341-f845-4bde-8190-a99accc40c96.png)
![dnp3_2](https://user-images.githubusercontent.com/84212412/126091961-f395855f-6c6d-41bc-8795-43dc780b4318.png)
[supervisord.conf.txt](https://github.com/riclolsen/json-scada/files/6837417/supervisord.conf.txt)

**Reply by riclolsen** [reply]
Yes.
It runs properly on Windows.
However, on Linux, it seems to have a problem with the build or with the Opendnp3 Nuget package.
I will try to fix that.

**Reply by riclolsen** [reply]
It is the opendnp3 Nuget package that is crashing on Linux.
However, it seems not possible to debug the package code on Linux to find the problem.
No solution yet.

**Reply by riclolsen** [reply]
I've just found this https://www.gitmemory.com/issue/dnp3/opendnp3/409/750781078 
It turns out it won't run on Linux because it uses Windows-only C++/CLI code on the package DLLs.
Sorry I wasn't aware of this library limitation.
Here https://github.com/dnp3/dnp3-simulator/issues/6 there is a suggestion to use PInvoke Dll Import. Not sure this is possible.
Probably I will have to rewrite this driver in C++ to make it fully portable.

**Reply by riclolsen** [reply]
I have a couple of suggestions as alternatives.
* As a console app this dnp3 driver can possibly run well under wine. You can bring the bin folder built for Windows and try to run Dnp3Client with Wine.
* You can run just the driver on a Windows computer or in a small Windows VM.

**Reply by rudyroc** [reply]
Thank you very much for the quick response and suggestions; I'll try these.
Any questions I will consult you. We will keep in touch.
Best regards,
Rudy

**Reply by rudyroc** [reply]
Hi Ricardo,
Good news !!.Very good suggestion. The Dnp3Client is running under wine properly. 
I will continue with the tests. Any news I will warn you.
Regards,
Rudy

**Reply by riclolsen** [reply]
I'm glad to know.
This is important information, please keep me posted with your results when possible.
Thank you very much, Rudy.

**Reply by riclolsen** [reply]
Hi Rudy,
I am trying to include the dnp3 driver on docker scripts using a Wine image. 
I can run the driver but it does not connect to MongoDB, also the mongo tool can not do it for some reason.
Did it run for you easily or it was required some special configuration of Wine?

**Reply by rudyroc** [reply]
Hi Ricardo
About your query, I have not made any special settings on wine.The wine version is 6.01.
I followed the following steps:
         I copied the bin folder of the windows version as bin_wine under the json-scada directory.
         I used supervisor to start the service DNP3 with the command line: 
                 command=wine ~/json-scada/bin_wine/Dnp3Client.exe 1 1
                 directory= ~/json-scada/bin_wine/
I hope this is helpful.
Best regards,
Rudy

**Comment by rudyroc** [comment]
Hi Ricardo
I tried to configure DNP3 to use the serial port but it didn't work.
I followed the instructions: leave blank the IP address and set the serial port to use. I tested both versions, Windows and Linux, and obtained the same result.
I attach Dnp3 Log output and configuration screen.
Best regards,
Rudy
![Captura de pantalla de 2021-09-01 01-14-00](https://user-images.githubusercontent.com/84212412/131779926-1e55d678-5141-4602-b6e8-1afb5c08ec25.png)
![Captura de pantalla de 2021-09-01 01-07-46](https://user-images.githubusercontent.com/84212412/131779930-7c480cca-4f8b-48a0-b69e-8e207acb787a.png)

**Reply by riclolsen** [reply]
Hi Rudy, there was a bad test that doesn't allow opening the serial port.
I have fixed it, please update the DNP3 driver code and recompile.
This fix will be also included in release 0.14 to be out soon.
Thank you for spotting the problem!

**Reply by rudyroc** [reply]
Hi Ricardo
Thanks for the new version. 
Now serial port connection is working correctly.
Best regards
